### ISSUE TITLE
[MOD: Firearms 3.0] Не работает спринт

### ISSUE THREAD

**Issue by Vladislav4KZ** [FWGS/xash3d-fwgs#1371]
На Xash3D, не работает спринт на клавише **X**.

К ней привязаны следующие команды:
```
+sprint;-sprint
```

На GoldSource, спринт начинает работать при нажатии и удержании клавиши **X**, и перестаёт работать по отпускании этой клавиши.

На Xash3D же, спринт не работает вообще.

Если на кнопку **X** забиндить команду ``+sprint`` то игрок будет бежать всегда после нажатия этой клавиши, вне зависимости от того, держите ли вы её или нет.

В Firearms3X, который по сути Firearms 3.0 работающий на Xash3D FWGS 0.20 образца 2020 года, в его README файле про этот баг тоже сказано.

А я проверял этот мод на самом последнем билде ксаша
```
v49/0.20 (win32-i386 build 2971)
```

Там этот баг всё ещё актуален.

**Comment by oscarsolerfollana** [comment]
I ran into this with Firearms and dug into it. Root cause confirmed by disassembling the mod's `client.dll` — it's an interaction between a mod bug and a deliberate engine difference:

**The mod bug:** `in_sprint.state` is missing from the reset block in `CL_ButtonBits` — every other button clears its "impulse down" bit (`state &= ~2`) after being read; sprint doesn't. That's why the mod ships `bind x "+sprint;-sprint"` instead of a plain `+sprint`: it's a workaround that happens to work on GoldSrc.

**The engine difference:** GoldSrc doesn't split key bindings on `;` — it appends the key number to the whole string on press, and flips the leading `+` to `-` on release. Xash3D splits the bind into commands and only passes the key number to the ones starting with `+` (`Key_AddKeyCommands`). So with the mod's default bind, Xash3D emits `-sprint` *without* an argument **on key press** — which `KeyUp` treats as a console-issued release and immediately clears the button. Result: sprint does nothing. And binding plain `+sprint` instead hits the mod bug and sprints forever, which matches both behaviors described in this issue.

The obvious fix (passing the key number to `-` commands too) makes it worse — that's the "sprints forever" case. What does reproduce GoldSrc is using GoldSrc's argument placement, but doing that unconditionally would re-break things like `bind x "+attack;+jump"`, which Xash3D handles better than GoldSrc precisely because it splits. A possible middle ground: use GoldSrc-style emission only when a bind uses the `+cmd;-cmd` idiom (starts with `+` and some segment after a `;` starts with `-`), and keep the current splitting for everything else. I have this written and tested with Firearms — sprint behaves exactly like on GoldSrc and normal binds are unaffected. It's a heuristic though, so I'd like to hear what you think before sending a PR.

**Comment by a1batross** [comment]
Thanks for your analysis!

This one of the cases where I think fixing the behavior for one mod, would break the assumed behavior for everybody else. To me, it's more of an accidental "feature" of GoldSrc.

Of course we can always introduce a new bugcomp mode (https://github.com/FWGS/xash3d-fwgs/blob/master/Documentation/bug-compatibility.md) but I wonder if there can be a solution that would satisfy all.

We can easily stick to GoldSrc behavior if this was present in original Xash3D and a regression by us OR if it was fixed in Xash3D, but none of Xash3D-based games rely on this behavior. Also it can be a genuine bug coming from Quake or Quake 2, meaning all Quake forks carried it by default.

**Comment by oscarsolerfollana** [comment]
Agreed on all points — unconditionally restoring GoldSrc behavior would just trade one set of broken binds for another (`bind x "+attack;+jump"` works on Xash precisely because of the splitting, and gets stuck on GoldSrc).

The middle ground I mentioned is narrower than a bugcomp mode and might satisfy all: use GoldSrc-style emission (key number appended to the whole string, leading `+` flipped to `-` on release) **only when the bind starts with `+` and some segment after a `;` starts with `-`** — the `+cmd;-cmd` idiom. Under the current behavior that idiom cannot work as intended for any mod: the `-cmd` half fires *at key press* with no key number, which either does nothing or (as in this case) immediately releases the button the `+cmd` half just pressed. So no existing setup can be relying on it, and every other bind keeps the current (better) splitting behavior untouched.

As for why the idiom exists at all: mods like Firearms ship it as a default bind to work around their own missing state reset (details above), and it happens to work on GoldSrc purely because of where GoldSrc places the key number. Whether it comes from Quake — Quake's `Key_Event` also appends the key number to the whole binding string, so yes, GoldSrc inherited that; the splitting is Xash's own (deliberate, and generally better) divergence.

I have this implemented and tested with Firearms: hold-to-sprint behaves exactly like on GoldSrc, and `+attack;+jump`-style binds are unaffected. Happy to send a PR if the approach sounds acceptable — or to gate it behind a bugcomp flag instead if you'd rather keep it opt-in.
